### PR TITLE
Refactoring of protocol logic to lower RAM consumption

### PR DESCRIPTION
**Pull request by DRracer**
Based on #59 by @gudnimg + the recent attempts to lower RAM usage and code size.

- [x] needs to be unit tested before merging

PFW-1395

Saves 200B of CPUFLASH. RAM usage dropped as well down to 5610B.

### CODE REVIEW

**Comment by gudnimg** [comment]
@DRracer It looks like the firmware version is not reported correctly:

image

My printer boots up with a FW mismatch error each time.

**Comment by DRracer** [comment]
@gudnimg it is reported correctly but it looks like the check is wrong ... `S0 A2` is correct

that's gonna be something with the pgm_read_* again :facepalm:

File: Firmware/mmu2_protocol_logic.cpp [hunk 1]
```diff
@@ -465,65 +461,65 @@ void ProtocolLogic::EjectFilament(uint8_t slot) {
...
+void ProtocolLogic::SwitchFromStartToIdle() {
     state = State::Running;
-    currentState = &idle;
-    idle.Restart();
-    idle.SendQuery(); // force sending Q0 immediately
-    idle.state = Idle::State::QuerySent;
+    currentScope = Scope::Idle;
+    IdleRestart();
+    SendQuery(); // force sending Q0 immediately
+    scopeState = ScopeState::QuerySent;
```
Comment (gudnimg) [inline Firmware/mmu2_protocol_logic.cpp]: This line is redundant. We set the state in `SendQuery()` (saves 6 bytes)

File: Firmware/mmu2_protocol_logic.cpp [hunk 2]
```diff
@@ -128,256 +124,253 @@ void ProtocolLogic::SendMsg(RequestMsg rq) {
...
+    case ScopeState::FilamentSensorStateSent:
         SendFINDAQuery();
-        break;
-    case State::FINDAReqSent:
-        return ProcessFINDAReqSent(Processing, State::Wait);
-    case State::ButtonSent:{
-        if (auto expmsg = logic->ExpectingMessage(linkLayerTimeout); expmsg != MessageReady)
-            return expmsg;
-        if (logic->rsp.paramCode == ResponseMsgParamCodes::Accepted) {
+        scopeState = ScopeState::FINDAReqSent;
```
Comment (gudnimg) [inline Firmware/mmu2_protocol_logic.cpp]: This line is redundant. We set the state in `SendFINDAQuery()` (saves 8 bytes)

File: Firmware/mmu2_protocol_logic.cpp [hunk 3]
```diff
@@ -388,17 +381,21 @@ StepStatus Idle::Step() {
         SendFINDAQuery();
         return Processing;
         break;
```
Comment (gudnimg) [inline Firmware/mmu2_protocol_logic.cpp]: Minor nitpick comment. But consider removing the break?

File: Firmware/mmu2_protocol_logic.cpp [hunk 4]
```diff
@@ -664,12 +664,12 @@ StepStatus ProtocolLogic::Step() {
         // We are ok, switching to Idle if there is no potential next request planned.
         // But the trouble is we must report a finished command if the previous command has just been finished
         // i.e. only try to find some planned command if we just finished the Idle cycle
-        bool previousCommandFinished = currentState == &command; // @@TODO this is a nasty hack :( 
-        if( ! ActivatePlannedRequest() ){ // if nothing is planned, switch to Idle
+        bool previousCommandFinished = currentScope == Scope::Command; // @@TODO this is a nasty hack :(
+        if (!ActivatePlannedRequest()) {                               // if nothing is planned, switch to Idle
             SwitchToIdle();
         } else {
```
Comment (gudnimg) [inline Firmware/mmu2_protocol_logic.cpp:667]: We don't want to call `SwitchToIdle()` if we are already in `Idle`. This causes the issue I mentioned on Slack.

`if(currentScope != Scope::Idle && !ActivatePlannedRequest())` seemed to work for me and only adds 6 bytes of flash.

File: Firmware/mmu2_protocol_logic.cpp [hunk 5]
```diff
@@ -128,256 +124,253 @@ void ProtocolLogic::SendMsg(RequestMsg rq) {
...
         }
-        break;
-    case State::S2Sent: // received response to S2 - revision
-        if( logic->rsp.request.code != RequestMsgCodes::Version || logic->rsp.request.value != 2 ){
+    }
+    return Processing;
+}
+
+StepStatus ProtocolLogic::ScopeStep() {
+    if ((uint_fast8_t)scopeState & (uint8_t)ScopeState::NotExpectsResponse) {
```
Comment (gudnimg) [inline Firmware/mmu2_protocol_logic.cpp]: I think this can be `if( !ExpectsResponse() )`, doesn't seem to save space but is more readable.

**Comment by gudnimg** [comment]
Another issue I'm seeing is if you boot up the printer with the MMU disconnected, the printer does not try to communicate with the MMU (No error).

EDIT: I'm wrong, the `MMU2_ECHO_MSG(tmp);` message is being hidden. `S0*99.` is sent repeatedly and no communication timeout occurs. Not sure if this is intended behavior.

File: Firmware/mmu2_protocol_logic.cpp (hunk 4 above)
Comment (DRracer) [reply Firmware/mmu2_protocol_logic.cpp:667]: thanks for pointing this out, fixed in 2c37f1db

**Comment by DRracer** [comment]
@gudnimg thank you for your detailed review :+1: , all your comments should be fixed now.

**Review by gudnimg — APPROVED** [review]
Awesome, I’m approving 💪

**Pull request merged**